File: netapp-manageability-sdk-9.3/src/sample/Data_ONTAP/C/nfs/nfs.c
//======================================================================//
//                                                                      //
// $ID$                                                                 //
//                                                                      //
// nfs.c                                                                //
//                                                                      //
// Brief information of the contents                                    //
//                                                                      //
// Copyright 2002-2003 Network Appliance, Inc. All rights               //
// reserved. Specifications subject to change without notice.           //
//                                                                      //
// This SDK sample code is provided AS IS, with no support or           //
// warranties of any kind, including but not limited to                 //
// warranties of merchantability or fitness of any kind,                //
// expressed or implied.  This code is subject to the license           //
// agreement that accompanies the SDK.                                  //
//                                                                      //
//  Sample for usage of following nfs group API:                        //
//                      nfs-enable                                      //
//                      nfs-disable                                     //
//                      nfs-status                                      //
//                      nfs-exportfs-list-rules                         //
//                                                                      //
// Usage:                                                               //
// nfs <filer> <user> <password> <operation>                            //
//                                                                      //
// <filer>      -- Name/IP address of the filer                         //
// <user>       -- User name                                            //
// <password>   -- Password                                             //
// <operation>  --                                                      //
//                 enable                                               //
//                 disable                                              //
//                 status                                               //
//                 list                                                 //
//======================================================================//


#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <netapp_api.h>


void print_usage() {
        printf("\nUsage : nfs <filer> <user> <passwd> <operation> \n");
        printf("<filer> -- ");
        printf("Name/IP address of the filer\n");
        printf("<user>  -- User Name\n");
        printf("<passwd>        -- Password\n");
        printf("<operation>     --\n");
        printf("\tenable - To enable NFS Service\n");
        printf("\tdisable - To disable NFS Service\n");
        printf("\tstatus - To print the status of NFS Service\n");
        printf("\tlist - To list the NFS export rules\n");
        exit (-1);
}

/****************************************************************************
 Name: Main

 Description: Initializes the server context object and calls the
              appropriate nfs API commands based on the input arguments

 Parameters:
      IN:
        argc    - number of command line arguments
        argv    - Array of command line arguments

 Return value: None
******************************************************************************/

int main(int argc, char* argv[])
{
        na_server_t*    s;
        na_elem_t       *out,*in,*export_info;
        na_elem_t*      nextElem = NULL;
        na_elem_t*      nextHost = NULL;
        na_elem_iter_t  iter, hostIter;
        char*           filerip    = argv[1];
        char*           user       = argv[2];
        char*           passwd     = argv[3];
        char*           operation  = argv[4];
        char            err[256];

        if (argc < 5) {
                print_usage();

        }

        // One-time initialization of system on client
        if (!na_startup(err, sizeof(err))) {
                fprintf(stderr, "Error in na_startup: %s\n", err);
                return -2;
        }

        //
        // Initialize connection to server, and request version 1.0 of the API set.
        //
        s = na_server_open(filerip, 1, 0);

        //
        // Set connection style (HTTP)
        //
        na_server_style(s, NA_STYLE_LOGIN_PASSWORD);
        na_server_set_transport_type(s,NA_SERVER_TRANSPORT_HTTP, NULL);
        na_server_adminuser(s,user,passwd);

	if(!strcmp(operation, "enable")) {


                in = na_elem_new("nfs-enable");
                out = na_server_invoke_elem(s,in);
                if (na_results_status(out) != NA_OK) {
                        printf("Error %d: %s\n", na_results_errno(out),
                                na_results_reason(out));
                        na_elem_free(out);
                        na_server_close(s);
                        na_shutdown();
                        return -3;
                }
				else {
					printf("enabled successfully!\n");
				}

        } else if(!strcmp(operation, "disable")) {


                in = na_elem_new("nfs-disable");
                out = na_server_invoke_elem(s,in);
                if (na_results_status(out) != NA_OK) {
                        printf("Error %d: %s\n", na_results_errno(out),
                                na_results_reason(out));
                        na_elem_free(out);
                        na_server_close(s);
                        na_shutdown();
                        return -3;
                }
				else {
					printf("disabled successfully!\n");
				}

        } else if(!strcmp(operation, "status")) {


                in = na_elem_new("nfs-status");
                out = na_server_invoke_elem(s,in);
                if (na_results_status(out) != NA_OK) {
                        printf("Error %d: %s\n", na_results_errno(out),
                                na_results_reason(out));
                        na_elem_free(out);
                        na_server_close(s);
                        na_shutdown();
                        return -3;
                }
                if (!strcmp(na_child_get_string(out,"is-enabled"),"true")) {
                        printf("NFS Server is enabled\n");
                } else {
                        printf("NFS Server is disabled\n");
                }

	 }else if(!strcmp(operation, "list")) {


                in = na_elem_new("nfs-exportfs-list-rules");

                out = na_server_invoke_elem(s,in);
                if (na_results_status(out) != NA_OK) {
                        printf("Error %d: %s\n", na_results_errno(out),
                                na_results_reason(out));
                        na_elem_free(out);
                        na_server_close(s);
                        na_shutdown();
                        return -3;
                }

                export_info = na_elem_child(out,"rules");
                if (export_info == NULL) {
                        // no data to list
                        na_elem_free(out);
                        na_server_close(s);
                        na_shutdown();
                        return -3;
                }
                // Iterate through retention date list
                for (iter=na_child_iterator(export_info); (nextElem =na_iterator_next(&iter)) != NULL;  ) {

                        char path_name[1024] = "";
                        char rw_list[1024] = "rw=";
                        char ro_list[1024] = "ro=";
                        char root_list[1024] = "root=";
                        char *host_name = NULL;
                        na_elem_t *results = NULL;

                        strcpy(path_name,na_child_get_string(nextElem,"pathname"));

                        if(na_elem_child(nextElem,"read-only") != NULL){
                                results = na_elem_child(nextElem,"read-only");

                                for (hostIter=na_child_iterator(results); (nextHost =na_iterator_next(&hostIter)) != NULL;  ) {

                                        if(na_child_get_string(nextHost,"all-hosts") != NULL){
                                                host_name = (char *)na_child_get_string(nextHost,"all-hosts");
                                                if(!strcmp(host_name,"true")){
                                                        strcat(ro_list,"all-hosts");
                                                        break;
                                                }
                                        } else if(na_child_get_string(nextHost,"name") != NULL) {
                                                host_name = (char *)na_child_get_string(nextHost,"name");
                                                strcat(ro_list,host_name);
                                                strcat(ro_list,":");
                                        }
                                }
                        }

			if(na_elem_child(nextElem,"read-write") != NULL){
                                results = na_elem_child(nextElem,"read-write");

                                for (hostIter=na_child_iterator(results); (nextHost =na_iterator_next(&hostIter)) != NULL;  ) {

                                        if(na_child_get_string(nextHost,"all-hosts") != NULL){
                                                host_name = (char *)na_child_get_string(nextHost,"all-hosts");
                                                if(!strcmp(host_name,"true")){
                                                        strcat(rw_list,"all-hosts");
                                                        break;
                                                }
                                        } else if(na_child_get_string(nextHost,"name") != NULL) {
                                                host_name = (char *)na_child_get_string(nextHost,"name");
                                                strcat(rw_list,host_name);
                                                strcat(rw_list,":");
                                        }
                                }
                        }
                        if(na_elem_child(nextElem,"root") != NULL){
                                results = na_elem_child(nextElem,"root");

                                for (hostIter=na_child_iterator(results); (nextHost =na_iterator_next(&hostIter)) != NULL;  ) {

                                        if(na_child_get_string(nextHost,"all-hosts") != NULL){
                                                host_name = (char *)na_child_get_string(nextHost,"all-hosts");
                                                if(!strcmp(host_name,"true")){
                                                        strcat(root_list,"all-hosts");
                                                        break;
                                                }
                                        } else if(na_child_get_string(nextHost,"name") != NULL) {
                                                host_name = (char *)na_child_get_string(nextHost,"name");
                                                strcat(root_list,host_name);
                                                strcat(root_list,":");
                                        }
                                }
                        }

                        strcat(path_name, "  ");
                        if(strcmp(ro_list,"ro=")) {
                                strcat(path_name,ro_list);
                        }
                        if(strcmp(rw_list,"rw=")) {
                                if(strstr(path_name,ro_list) != NULL) {
                                        strcat(path_name,",");
                                }
                                strcat(path_name,rw_list);
                        }
                        if(strcmp(root_list,"root=")) {
                                if((strstr(path_name,ro_list) != NULL) || (strstr(path_name,rw_list) != NULL)) {
                                        strcat(path_name,",");
                                }
                                strcat(path_name,root_list);
                        }


                        printf("%s\n",path_name);
                }

 	} else {
                printf("Invalid Operation \n");
                print_usage();
        }

        na_elem_free(out);
        na_server_close(s);
        na_shutdown();

        return 0;

}

File: netapp-manageability-sdk-9.3/src/sample/Data_ONTAP/Ruby/nfs.rb
#=======================================================================#
#                                                                       #
# $ID$                                                                  #
#                                                                       #
# nfs.rb                                                                #
#                                                                       #
# Sample code for the following APIs:                                   #
#               nfs-enable, nfs-disable                                 #
#               nfs-status, nfs-exportfs-list-rules,                    #
#                                                                       #
# Copyright 2011 Network Appliance, Inc. All rights                     #
# reserved. Specifications subject to change without notice.            #
#                                                                       #
# This SDK sample code is provided AS IS, with no support or            #
# warranties of any kind, including but not limited to                  #
# warranties of merchantability or fitness of any kind,                 #
# expressed or implied.  This code is subject to the license            #
# agreement that accompanies the SDK.                                   #
#=======================================================================#

$:.unshift '../../../../lib/ruby/NetApp'
require 'NaServer'

def print_usage
    print ("Usage:\n")
    print ("nfs.rb <storage> <user> <password> <command>\n")
    print ("<storage> -- storage system name\n")
    print ("<user> -- User name\n")
    print ("<password> -- Password\n")
    print ("<command> -- enable, disable, status, list\n")
    exit
end


args = ARGV.length
if(args < 4) 
    print_usage()
end
$storage = ARGV[0]
$user = ARGV[1]
$pw = ARGV[2]
$cmd = ARGV[3]


def do_nfs()
    s = NaServer.new($storage, 1, 3)
    out = s.set_admin_user($user, $pw)
    if($cmd == "enable")
        out = s.invoke("nfs-enable")
        if(out.results_status() == "failed")
            print(out.results_reason() + "\n")
            exit		
        else
            print("Operation successful\n")
        end		
    elsif ($cmd == "disable")
        out = s.invoke("nfs-disable")
        if(out.results_status == "failed")
            print(out.results_reason() + "\n")
            exit
	else
            print("Operation successful\n")
	end		
    elsif ($cmd == "status")
        out = s.invoke("nfs-status")
        if(out.results_status == "failed")
            print(out.results_reason() + "\n")
            exit
	end
	enabled = out.child_get_string("is-enabled")
        if(enabled == "true")
            print("NFS Server is enabled\n")
        else
            print("NFS Server is disabled\n")
	end		
    elsif ( $cmd == "list") 
        out = s.invoke( "nfs-exportfs-list-rules" )
        export_info = out.child_get("rules")
        result = export_info.children_get()
        result.each do |export|
            path_name = export.child_get_string("pathname")
            rw_list = "rw="
            ro_list = "ro="
            root_list = "root="			
            if(export.child_get("read-only"))
                ro_results = export.child_get("read-only")
                ro_hosts = ro_results.children_get()
		ro_hosts.each do |ro|
                    if(ro.child_get_string("all-hosts"))
                        all_hosts = ro.child_get_string("all-hosts")
                        if(all_hosts == "true") 
                            ro_list = ro_list + "all-hosts"
                            break
			end
                    elsif(ro.child_get_string("name")) 
                        host_name = ro.child_get_string("name")
                        ro_list = ro_list + host_name + ":"
		    end
		end
	    end			
            if(export.child_get("read-write"))
                rw_results = export.child_get("read-write")
                rw_hosts = rw_results.children_get()                
                rw_hosts.each do |rw|
                    if(rw.child_get_string("all-hosts"))
                        all_hosts = rw.child_get_string("all-hosts")
                        if(all_hosts == "true") 
                            rw_list = rw_list + "all-hosts"
                            break
			end						
                    elsif(rw.child_get_string("name"))
                        host_name = rw.child_get_string("name")
                        rw_list = rw_list + host_name + ":"
	            end
		end
	    end			
            if(export.child_get("root"))
                root_results = export.child_get("root")
                root_hosts = root_results.children_get()
		root_hosts.each do |root|
                    if(root.child_get_string("all-hosts"))
                        all_hosts = root.child_get_string("all-hosts")
                        if(all_hosts == "true")
                            root_list = root_list + "all-hosts"
                            break
			end
                    elsif(root.child_get_string("name"))
                        host_name = root.child_get_string("name")
                        root_list = root_list + host_name + ":"
		    end
		end
	    end
            path_name = path_name + "  "
            if(ro_list != "ro=") 
		path_name = path_name + ro_list 
	    end
            if(rw_list != "rw=") 
		path_name = path_name + "," + rw_list
	    end
            if(root_list != "root=")
		path_name = path_name + "," + root_list 
	    end
            print (path_name + "\n")
	end
    else 
        print("Invalid operation\n")
        print_usage()
    end
end


do_nfs()

